Client/WebFilters/ShieldStats/Adblock/AdBlockEngine+Extensions.swift

```diff
@@ -0,0 +1,153 @@
+    let groupedResources = group(resources: resources)
+    
+    let enginesWithCompileResults = await groupedResources.asyncConcurrentMap { source, resources -> (engine: CachedAdBlockEngine, compileResults: [AdBlockEngineManager.ResourceWithVersion: Result<Void, Error>]) in
+      let results = await CachedAdBlockEngine.createEngine(from: resources, source: source)
+      return (results.engine, results.compileResults)
+    }
+    
+    var allCompileResults: [AdBlockEngineManager.ResourceWithVersion: Result<Void, Error>] = [:]
+    let engines = enginesWithCompileResults.map({ $0.engine })
+    
+    for result in enginesWithCompileResults {
+      for compileResult in result.1 {
+        allCompileResults[compileResult.key] = compileResult.value
+      }
+    }
+    
+    return (engines, allCompileResults)
+  }
+  
+  /// Create an engine from the given resources.
+  ///
+  /// - Warning: You should only have at max dat file in this list. Each dat file can only represent a single engine
+  public static func createEngine(from resources: [AdBlockEngineManager.ResourceWithVersion]) -> (engine: AdblockEngine, compileResults: [AdBlockEngineManager.ResourceWithVersion: Result<Void, Error>]) {
+    let combinedRuleLists = Self.combineAllRuleLists(from: resources)
+    // Create an engine with the combined rule lists
+    let engine = AdblockEngine(rules: combinedRuleLists)
+    // Compile remaining resources
+    let compileResults = engine.compile(resources: resources)
+    // Return the compiled data
+    return (engine, compileResults)
+  }
+  
+  /// Combine all resources of type rule lists to one single string
+  private static func combineAllRuleLists(from resourcesWithVersion: [AdBlockEngineManager.ResourceWithVersion]) -> String {
+    // Combine all rule lists that need to be injected during initialization
+    let allResults = resourcesWithVersion.compactMap { resourceWithVersion -> String? in
+      switch resourceWithVersion.resource.type {
+      case .ruleList:
+        guard let data = FileManager.default.contents(atPath: resourceWithVersion.fileURL.path) else {
+          return nil
+        }
+        
+        return String(data: data, encoding: .utf8)
+      case .dat, .jsonResources:
+        return nil
+      }
+    }
+    
+    let combinedRules = allResults.joined(separator: "\n")
+    return combinedRules
+  }
+  
+  static func group(
+    resources: [AdBlockEngineManager.ResourceWithVersion]
+  ) -> [AdBlockEngineManager.Source: [AdBlockEngineManager.ResourceWithVersion]] {
+    var groups: [AdBlockEngineManager.Source: [AdBlockEngineManager.ResourceWithVersion]] = [:]
+    
+    for resourceWithVersion in resources {
+      var group = groups[resourceWithVersion.resource.source] ?? []
+      group.append(resourceWithVersion)
+      groups[resourceWithVersion.resource.source] = group
+    }
+    
+    return groups
```

Comment:
Just use `Dictionary(grouping:by:)` instead:

```swift
return Dictionary(grouping: resources, by: \.resource.source)
```

Also for future ref, you can use the `default` subscript argument to avoid the pattern inside the loop:

```swift
groups[resourceWithVersion.resource.source, default: []].append(resourceWithVersion)
```